tests/unit/client-eviction.tcl

```diff
@@ -0,0 +1,475 @@
+        error "no client named $name found with field $f"
+    }
+    return $res
+}
+
+# Sum a value across all redis client connections:
+# f - the field name from "CLIENT LIST" we want to sum
+proc clients_sum {f} {
+    set sum 0
+    set clients [split [string trim [r client list]] "\r\n"]
+    foreach c $clients {
+        if {![regexp $f=(\[a-zA-Z0-9-\]+) $c - res]} {
+            error "field $f not found in $c"
+        }
+        incr sum $res
+    }
+    return $sum
+}
+
+proc write_err_exception {e} {
+    return [regexp {(.*connection reset by peer.*|.*broken pipe.*|error writing .*: protocol wrong type for socket)} $e]
+}
+
+proc mb {v} {
+    return [expr $v * 1024 * 1024]
+}
+
+start_server {} {
+    set maxmemory_clients 3000000
+    r config set maxmemory-clients $maxmemory_clients
+    
+    test "client evicted due to large argv" {
+        r flushdb
+        set rr [redis_client]
+        # Attempt a large multi-bulk command under eviction limit
+        $rr mset k v k2 [string repeat v 1000000]
+        assert_equal [$rr get k] v
+        # Attempt another command, now causing client eviction
+        catch { $rr mset k v k2 [string repeat v $maxmemory_clients] } e
+        assert {[write_err_exception $e]}
+    }
+
+    test "client evicted due to large query buf" {
+        r flushdb
+        set rr [redis_client]
+        # Attempt to fill the query buff without completing the argument above the limit, causing client eviction
+        catch { 
+            $rr write [join [list "*1\r\n\$$maxmemory_clients\r\n" [string repeat v $maxmemory_clients]] ""]
+            $rr flush
+            $rr read
+        } e
+        assert {[write_err_exception $e]}
+    }
+
+    test "client evicted due to percentage of maxmemory" {
+        set maxmemory [mb 6]
+        r config set maxmemory $maxmemory
+        # Set client eviction threshold to 7% of maxmemory        
+        set maxmemory_clients_p 7
+        r config set maxmemory-clients $maxmemory_clients_p%
+        r flushdb
+        
+        set maxmemory_clients_actual [expr $maxmemory * $maxmemory_clients_p / 100]
+
+        set rr [redis_client]
+        $rr client setname test_client
+        # Attempt to fill the query buff with only half the percentage threshold verify we're not disconnected
+        set n [expr $maxmemory_clients_actual / 2]
+        $rr write [join [list "*1\r\n\$$n\r\n" [string repeat v $n]] ""]
+        $rr flush
+        set tot_mem [client_field test_client tot-mem]
+        assert {$tot_mem >= $n && $tot_mem < $maxmemory_clients_actual}
+        
+        # Attempt to fill the query buff with the percentage threshold of maxmemory and verify we're evicted
+        $rr close
+        set rr [redis_client]
+        catch { 
+            $rr write [join [list "*1\r\n\$$maxmemory_clients_actual\r\n" [string repeat v $maxmemory_clients_actual]] ""]
+            $rr flush
+        } e
+        assert {[write_err_exception $e]}
```

Comment:
I run test: `./runtest --single unit/client-eviction --stop --loop --only "client evicted due to percentage of maxmemory"`
This test fails easily(whether it's macos or ubuntu), I print out `$e` which is empty, then when I try to add more small strings of wirte and flush operations after it, the test doesn't fail.
```suggestion
        catch { 
            $rr write [join [list "*1\r\n\$$maxmemory_clients_actual\r\n" [string repeat v $maxmemory_clients_actual]] ""]
            $rr flush
            
            for {set j 0} {$j < 100} {incr j} {
                after 1
                $rr write [join [list "*1\r\n\$1\r\n" "v"] ""]
                $rr flush
            }
        } e
        assert {[write_err_exception $e]}
```

Reply:
Thanks, I see the issue. 
I wonder why it fails, the `flush` should verify the write completed and the write does fail, I see it evicted the client in the logs. I'll investigate a bit since I don't want to add the extra writes without understanding this...

Reply:
I changed most of the tests to just verify the client was evicted as expected instead of detecting a socket error. This is more robust as I saw in the logs that everything actually works fine it's just that the `write` somehow succeeds without throwing an error even though the server does the eviction.
Anyway it's much more stable now. 
If you find anything else let me know..